gluon improvement

szha opened the pull request:
changes include:
1. use generator expr for arguments when calling cached op
2. merge two separate for loops for assembling parameters into one so that the check for parameter membership in `cargs` is cached.
3. fill shapes after parameter data is set, to be reflected in block's `__repr__`.

piiswrong commented:
Construction is only called once

piiswrong commented on python/mxnet/gluon/block.py:
```diff
@@ -384,32 +384,33 @@ def _build_cache(self, *args):
-        assert fmt == self._in_format, "Invalid input format"
-        for i, j in self._in_idx:
-            cargs[i] = args[j]
+            cargs = tuple(args[self._in_idx[i]] if self._in_idx[i] != -1 else v for i, v
```
This doesn't look right

szha replied:
This is a simple combination of the list comprehension and the for loop. What seems to be the problem?

szha replied:
All tests passed.

Could you double check and see if still doesn't look right? If so, what looks like the error? And what other tests should be added to catch the error?

piiswrong commented:
What's the benefit of this change? This part is complicated and hard to debug, so unless there is visible performance impact I think there is no reason to change it

mli commented:
Agree that using tuple and iterator can reduce the memory. But Eric's point is valid, this PR puts too many things in a single expression (i.e. `cargs = ...`), which makes the codes harder to read. 

maybe we can write a generator for cargs, which can further reduce memory, and break single line into multiple lines to improve readability. 

also, is there any benchmark showing that this kinds of modification can improve things in a notiable way?

szha commented:
Thanks for the comments, I worked on this a bit more and simplified the assembly of calling arguments using one loop. I renamed the field for readability. And here are the timing results:

workload:
```python
import mxnet as mx
from mxnet import gluon
class example(gluon.HybridBlock):
    def __init__(self, num_args, num_params):
        super(example, self).__init__()
        self._num_args = num_args
        self._num_params = num_params
        for i in range(num_params):
            setattr(self, 'param%d'%i, self.params.get('param%d'%i, shape=(1,), init='ones'))
    def hybrid_forward(self, F, *args, **kwargs):
        return F.concat(*(a for a in (args+tuple(kwargs.values()))), dim=0)

num_args, num_params = 2, 50
net = example(num_args, num_params)
net.initialize()
net.hybridize()
```

Results:
```
num_args, num_params = 100, 200
100 loops, best of 3: 5.52 ms per loop
100 loops, best of 3: 5.37 ms per loop

num_args, num_params = 2, 50
100 loops, best of 3: 333 µs per loop
100 loops, best of 3: 312 µs per loop
```

piiswrong commented:
Check pointing is not a bottleneck and collecting param is unlikely a big cost during checkpointing.

This is over complicated. Optimization should be driven by profiling. Premature optimization is the source of all evil.

When a layer is changed, simply clear cached op.

szha commented:
Reverted per request and ready for review.

szha commented:
added shape completion after loading parameters or finishing deferred init.

cjolivier01 commented:
Premature optimization is the root of ALL evil? Wow. :)

szha commented:
yeah, thou shall yield when Knuth is quoted.

cjolivier01 commented:
Well, this certainly explains the Vegas Massacre!

piiswrong commented on python/mxnet/gluon/rnn/rnn_layer.py:
```diff
@@ -89,8 +89,9 @@ def __repr__(self):
-        mapping = ('{_input_size} -> {_hidden_size}'.format(**self.__dict__) if self._input_size
-                   else self._hidden_size)
+        shape = self.i2h_weight[0].shape
+        mapping = ('{0} -> {1}'.format(shape[1], shape[0]) if shape[1]
```
It should show None -> xx when shape is not defined yet
